core/services/directrequestocr/models.go

```diff
@@ -3,21 +3,123 @@ package directrequestocr
 import (
 	"database/sql/driver"
 	"encoding/hex"
+	"errors"
 	"fmt"
 	"time"
 
 	"github.com/ethereum/go-ethereum/common"
 )
 
+const RequestIDLength int = 32
+
+type RequestID [RequestIDLength]byte
+type Request struct {
+	RequestID         RequestID
+	RunID             *int64
+	ReceivedAt        time.Time
+	RequestTxHash     *common.Hash
+	State             RequestState
+	ResultReadyAt     *time.Time
+	Result            []byte
+	ErrorType         *ErrType
+	Error             []byte
+	TransmittedResult []byte
+	TransmittedError  []byte
+	// TODO: add timestamps for other possible states: https://app.shortcut.com/chainlinklabs/story/58428/timestamp-fields-for-all-states-in-ocr2dr-data-model
+}
+
 type RequestState int8
 
 const (
+	// IN_PROGRESS is the initial state of a request, set right after receiving it in an on-chain event.
 	IN_PROGRESS RequestState = iota
+
+	// RESULT_READY means that computation has finished executing (with either success or error).
+	// OCR2 reporting includes only requests in RESULT_READY state (for Query and Observation phases).
 	RESULT_READY
+
+	// TIMED_OUT request has not been transmitted yet and has been waiting too long for OCR2 processing.
+	// It won't be included in OCR2 reporting rounds any more.
+	TIMED_OUT
+
+	// TRANSMITTED request is one that passed through ShouldTransmitAcceptedReport phase of OCR2 reporting.
+	// From here, it can only transition to CONFIRMED state, upon receiving an on-chain confirmation from the oracle contract.
 	TRANSMITTED
+
+	// CONFIRMED state indicates that we received an on-chain confirmation event
+	// (with or without this node's participation in an earlier OCR round).
+	// We can transition here at any time (full fan-in) and cannot transition out (empty fan-out).
+	// This is a desired and expected final state for every request.
 	CONFIRMED
 )
 
+/*
+ *    +-----------+
+ * +--+IN_PROGRESS+----------------+
+ * |  +-----+-----+                |
+ * |        |                      |
+ * |        v                      v
+ * |  +------------+          +---------+
+ * |  |RESULT_READY+--------->|TIMED_OUT|
+ * |  +-----+------+          +---------+
+ * |        |
+ * |        v
+ * |  +-----------+
+ * +->|TRANSMITTED|
+ *    +-----------+
+ *
+ *                     \   /
+ *                       |
+ *                       v
+ *                  +---------+
+ *                  |CONFIRMED|
+ *                  +---------+
+ */
+func CheckStateTransition(prev RequestState, next RequestState) error {
+	if prev == next {
```

Comment:
You could inject these two cases/errors into your maps of maps. E.g. create error instances once, reuse them in the maps.
This would keep all declarations clear in one place (structure).